File: perl/src/parser.pl
#!/usr/bin/perl
#Prototype parsing code. Accepts one file as input and prints organized metadata in the form of an html post.
#The file parsed must be a plaintext file produced from mid3v2. 
#-Danny Jones

use Fcntl;
use Song; #see Song.pm
use Getopt::Long;

GetOptions ('ogg' => \$ogg);
GetOptions ('mp3' => \$mp3);
if ($#ARGV +1 != 1)
{
	print "Please specify one file to read. (EX. ./parser.pl filename.txt )\n";
	exit 0;
}
my $filename = "PLAYLIST";
my $filepath = $ARGV[0];
my $currentSong = new Song();
my @playlist = ();

#parsing functions
sub parseTitle 
{
	my $title = substr $_[0], 5;
	chomp $title;
	$currentSong->setTitle( $title );	
}

sub parseGenre
{
	my $genre = substr $_[0], 5;
	chomp $genre;
	$currentSong->setGenre( $genre );	
}

sub parseArtist
{
	my $artist = substr $_[0], 5;
	chomp $artist;
	$currentSong->setArtist( $artist );	
}

sub parseAlbum
{
	my $album = substr $_[0], 5;
	chomp $album;
	$currentSong->setAlbum( $album );	
}
sub parseLicense
{
	my $line = substr $_[0], 5;
	my $license = "";
	#Detects license components in the URL.
	if($line =~ m/creativecommons.org/g)
	{
	
		if($line =~ m/\/by\//g)
		{
			$license = "CC BY";
		}
		if($line =~ m/\/by-sa\//g)
		{
			$license = "CC BY-SA";
		}
		if($line =~ m/\/by-nc\//g)
		{
			$license = "CC BY-NC";
		}
		if($line =~m/\/by-nc-sa\//g)
		{
			$license = "CC BY-NC-SA";
		}
	}
	chomp $license;
	$currentSong->setLicense( $license );	

}
#FMA format
sub parseURL 
{
	my $line = substr ( $_[0], index($_[0], "URL:") + 5);
	chomp $line;
	$currentSong->setSource( $line );
}
#Jamendo Format
sub parseSource
{
	my $source = substr $_[0], 5;
	chomp $source;
	$currentSong->setSource( $source);	
}


open($filename, $filepath) or die "Cannot open file";

#print "Parsing $filepath...\n\n";
sub parseMP3 
{
	while(<$filename>)
	{
		if($_ =~ m/TIT2/g) 
		{
			parseTitle $_;
		}
		if($_ =~ m/TCON/g)
		{
			parseGenre $_;
		}
		if($_ =~ m/TPE1/g)
		{
			parseArtist $_;
		}
		if($_ =~ m/TALB/g)
		{
			parseAlbum $_;
		}
		if($_ =~ m/TCOP/g)
		{
			parseLicense $_;
		}
		if($_ =~ m/URL:/g)
		{
			parseURL $_;
		}
		if($_ =~ m/WOAS/g)
		{
			parseSource $_;
		}
		if($_ =~ m/SONG END/g) 
		{
			push @playlist, $currentSong;
			$currentSong = new Song();
		}
	}
}
sub parseOgg
{
}
if($ogg)
{
	parseOgg;
}
else
{
	parseMP3;
}
	
print "<br />\n<ol>\n";
#foreach our $song(@playlist)
#{
#	$song->printSong();
#}
foreach our $song(@playlist)
{
	$song->printSongHTML();
}
print "</ol><br />\n";
close($filename);

File: python/metadata_parser/metadata/parser.py
# vim: tabstop=4 shiftwidth=4 softtabstop=4

#!/usr/bin/env python

""" This module contains classes and utilities for parsing the output of the
mid3v2 and vorbiscomment utilities, as well as any websites provided.

The parser stores the information in an array of Song objects.
"""

from metadata.song import Song


class MIDParser(object):
    TITLE_MATCH = "TIT2"
    GENRE_MATCH = "TCON"
    ARTIST_MATCH = "TPE1"
    ALBUM_MATCH = "TALB"
    LICENSE_MATCH = "TCOP"
    SOURCE_MATCH_FMA = "URL:"
    SOURCE_MATCH_JAMENDO = "WOAS"

    def __init__(self, filepath=""):
        self.filepath = filepath
        self.currentsong = Song()
        self.playlist = []
        self.webparser = None

    def parse(self):
        for line in open(self.filepath):
            if self.TITLE_MATCH in line:
                self.parse_title(line)
            if self.GENRE_MATCH in line:
                self.parse_genre(line)
            if self.ARTIST_MATCH in line:
                self.parse_artist(line)
            if self.ALBUM_MATCH in line:
                self.parse_album(line)
            if self.LICENSE_MATCH in line:
                self.parse_license(line)
            if self.SOURCE_MATCH_FMA in line:
                self.parse_url_fma(line)
            if self.SOURCE_MATCH_JAMENDO in line:
                self.parse_url_jamendo(line)
            if "SONG END" in line:
                #TODO: Add functionality for WebParser to extract BandPage
                self.playlist.append(self.currentsong)
                self.currentsong = Song()

    def parse_title(self, line):
        title = line[5:].strip()
        self.currentsong.title = title

    def parse_genre(self, line):
        genre = line[5:].strip()
        self.currentsong.genre = genre

    def parse_artist(self, line):
        artist = line[5:].strip()
        self.currentsong.artist = artist

    def parse_album(self, line):
        album = line[5:].strip()
        self.currentsong.album = album

    def parse_license(self, line):
        line = line[5:].strip()
        song_license = ""
        if "creativecommons.org" in line:
            if "/by/" in line:
                song_license = "CC BY"
            if "/by-sa/" in line:
                song_license = "CC BY-SA"
            if "/by-nc/" in line:
                song_license = "CC BY-NC"
            if "/by-nc-sa/" in line:
                song_license = "CC BY-NC-SA"
        self.currentsong.license = song_license

    def parse_url_fma(self, line):
        url_index = line.find("URL:") + 5
        source = line[url_index:].strip()
        self.currentsong.source = source

    def parse_url_jamendo(self, line):
        source = line[5:].strip()
        self.currentsong.source = source

    def clear_songs(self):
        self.playlist = []


class VorbisParser(MIDParser):
    TITLE_MATCH = "TITLE"
    GENRE_MATCH = "GENRE"
    ARTIST_MATCH = "ALBUMARTIST"
    ALBUM_MATCH = "ALBUM"
    LICENSE_MATCH = "TCOP"

    def parse_title(self, line):
        title = line[6:].strip()
        self.currentsong.title = title

    def parse_genre(self, line):
        genre = line[6:].strip()
        self.currentsong.genre = genre

    def parse_artist(self, line):
        artist = line[12:].strip()
        self.currentsong.artist = artist

    def parse_album(self, line):
        album = line[6:].strip()
        self.currentsong.album = album
